### PR TITLE
Add System.Type converter for JsonSerializer

### LINKED ISSUE #31567: {De}Serializing System.Type to JSON using JsonSerializer isn't supported and fails with cycle detection (JsonException)
Scenario from https://github.com/dotnet/corefx/issues/41002#issuecomment-558058956 by @doogdeb

Serializing/deserializing a POCO containing a `System.Type` properties doesn't work today and fails with JsonException (A possible object cycle was detected which is not supported...). The workaround is to use a custom converter.

```C#
public static void SerializePocoWithType()
{
    var cache = new CacheModel
    {
        DtoModelType = typeof(string),
        CacheKey = "ABC"
    };

    // {"DtoModelType":"System.String, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e","CacheKey":"ABC"}
    Console.WriteLine(JsonConvert.SerializeObject(cache));
    Console.WriteLine(JsonConvert.SerializeObject(JsonConvert.DeserializeObject<CacheModel>(JsonConvert.SerializeObject(cache))));

    // Workaround for System.Type with S.T.Json using converters
    var options = new JsonSerializerOptions();
    options.Converters.Add(new CustomJsonConverterForType());
    Console.WriteLine(JsonSerializer.Serialize(cache, options));
    Console.WriteLine(JsonSerializer.Serialize(JsonSerializer.Deserialize<CacheModel>(JsonSerializer.Serialize(cache, options), options), options));

    JsonSerializer.Serialize(cache);
    /*
        * System.Text.Json.Serialization.Tests.ReadValueTests.SerializePocoWithType [FAIL]
System.Text.Json.JsonException : A possible object cycle was detected which is not supported. This can either be due to a cycle or if the object depth is larger than the maximum allowed depth of 0.
Stack Trace:
E:\GitHub\Fork\corefx\src\System.Text.Json\src\System\Text\Json\ThrowHelper.Serialization.cs(33,0): at System.Text.Json.ThrowHelper.ThrowInvalidOperationException_SerializerCycleDetected(Int32 maxDepth)
E:\GitHub\Fork\corefx\src\System.Text.Json\src\System\Text\Json\Serialization\JsonSerializer.Write.cs(45,0): at System.Text.Json.JsonSerializer.Write(Utf8JsonWriter writer, Int32 originalWriterDepth, Int32 flushThreshold, JsonSerializerOptions options, WriteStack& state)
E:\GitHub\Fork\corefx\src\System.Text.Json\src\System\Text\Json\Serialization\JsonSerializer.Write.Helpers.cs(140,0): at System.Text.Json.JsonSerializer.WriteCore(Utf8JsonWriter writer, Object value, Type type, JsonSerializerOptions options)
E:\GitHub\Fork\corefx\src\System.Text.Json\src\System\Text\Json\Serialization\JsonSerializer.Write.Helpers.cs(112,0): at System.Text.Json.JsonSerializer.WriteCore(PooledByteBufferWriter output, Object value, Type type, JsonSerializerOptions options)
E:\GitHub\Fork\corefx\src\System.Text.Json\src\System\Text\Json\Serialization\JsonSerializer.Write.Helpers.cs(87,0): at System.Text.Json.JsonSerializer.WriteCoreString(Object value, Type type, JsonSerializerOptions options)
E:\GitHub\Fork\corefx\src\System.Text.Json\src\System\Text\Json\Serialization\JsonSerializer.Write.String.cs(21,0): at System.Text.Json.JsonSerializer.Serialize[TValue](TValue value, JsonSerializerOptions options)
E:\GitHub\Fork\corefx\src\System.Text.Json\tests\Serialization\ReadValueTests.cs(29,0): at System.Text.Json.Serialization.Tests.ReadValueTests.SerializePocoWithType()
        */

    JsonSerializer.Serialize(cache, new JsonSerializerOptions { MaxDepth = 1_000 });
    /*
        * SerializePocoWithType [FAIL]
System.Text.Json.JsonException : The object or value could not be serialized. Path: $.DtoModelType.Assembly.DefinedTypes.GenericTypeParameters.Assembly.DefinedTypes.GenericTypeParameters.Assembly....DefinedTypes.GenericTypeParameters.Assembly.DefinedTypes.GenericTypeParameters.
---- System.InvalidOperationException : CurrentDepth (1000) is equal to or larger than the maximum allowed depth of 1000. Cannot write the next JSON object or array.
Stack Trace:
E:\GitHub\Fork\corefx\src\System.Text.Json\src\System\Text\Json\ThrowHelper.Serialization.cs(202,0): at System.Text.Json.ThrowHelper.ReThrowWithPath(WriteStack& writeStack, Exception ex)
E:\GitHub\Fork\corefx\src\System.Text.Json\src\System\Text\Json\Serialization\JsonSerializer.Write.cs(70,0): at System.Text.Json.JsonSerializer.Write(Utf8JsonWriter writer, Int32 originalWriterDepth, Int32 flushThreshold, JsonSerializerOptions options, WriteStack& state)
E:\GitHub\Fork\corefx\src\System.Text.Json\src\System\Text\Json\Serialization\JsonSerializer.Write.Helpers.cs(140,0): at System.Text.Json.JsonSerializer.WriteCore(Utf8JsonWriter writer, Object value, Type type, JsonSerializerOptions options)
E:\GitHub\Fork\corefx\src\System.Text.Json\src\System\Text\Json\Serialization\JsonSerializer.Write.Helpers.cs(112,0): at System.Text.Json.JsonSerializer.WriteCore(PooledByteBufferWriter output, Object value, Type type, JsonSerializerOptions options)
E:\GitHub\Fork\corefx\src\System.Text.Json\src\System\Text\Json\Serialization\JsonSerializer.Write.Helpers.cs(87,0): at System.Text.Json.JsonSerializer.WriteCoreString(Object value, Type type, JsonSerializerOptions options)
E:\GitHub\Fork\corefx\src\System.Text.Json\src\System\Text\Json\Serialization\JsonSerializer.Write.String.cs(21,0): at System.Text.Json.JsonSerializer.Serialize[TValue](TValue value, JsonSerializerOptions options)
E:\GitHub\Fork\corefx\src\System.Text.Json\tests\Serialization\ReadValueTests.cs(30,0): at System.Text.Json.Serialization.Tests.ReadValueTests.SerializePocoWithType()
----- Inner Stack Trace -----
E:\GitHub\Fork\corefx\src\System.Text.Json\src\System\Text\Json\ThrowHelper.cs(142,0): at System.Text.Json.ThrowHelper.ThrowInvalidOperationException(String message)
E:\GitHub\Fork\corefx\src\System.Text.Json\src\System\Text\Json\ThrowHelper.cs(138,0): at System.Text.Json.ThrowHelper.ThrowInvalidOperationException(Int32 currentDepth)
E:\GitHub\Fork\corefx\src\System.Text.Json\src\System\Text\Json\Writer\Utf8JsonWriter.cs(606,0): at System.Text.Json.Utf8JsonWriter.WriteStartHelper(ReadOnlySpan`1 utf8PropertyName, Byte token)
E:\GitHub\Fork\corefx\src\System.Text.Json\src\System\Text\Json\Writer\Utf8JsonWriter.cs(583,0): at System.Text.Json.Utf8JsonWriter.WriteStartArray(JsonEncodedText propertyName)
E:\GitHub\Fork\corefx\src\System.Text.Json\src\System\Text\Json\Serialization\WriteStackFrame.cs(90,0): at System.Text.Json.WriteStackFrame.WriteObjectOrArrayStart(ClassType classType, JsonEncodedText propertyName, Utf8JsonWriter writer, Boolean writeNull)
E:\GitHub\Fork\corefx\src\System.Text.Json\src\System\Text\Json\Serialization\WriteStackFrame.cs(72,0): at System.Text.Json.WriteStackFrame.WriteObjectOrArrayStart(ClassType classType, Utf8JsonWriter writer, JsonSerializerOptions options, Boolean writeNull)
E:\GitHub\Fork\corefx\src\System.Text.Json\src\System\Text\Json\Serialization\JsonSerializer.Write.HandleEnumerable.cs(47,0): at System.Text.Json.JsonSerializer.HandleEnumerable(JsonClassInfo elementClassInfo, JsonSerializerOptions options, Utf8JsonWriter writer, WriteStack& state)
E:\GitHub\Fork\corefx\src\System.Text.Json\src\System\Text\Json\Serialization\JsonSerializer.Write.cs(45,0): at System.Text.Json.JsonSerializer.Write(Utf8JsonWriter writer, Int32 originalWriterDepth, Int32 flushThreshold, JsonSerializerOptions options, WriteStack& state)
Finished:    System.Text.Json.Test
        */
}

// Sample workaround for System.Type with S.T.Json using converters
internal class CustomJsonConverterForType : JsonConverter<Type>
{
    public override Type Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        // Caution: Deserialization of type instances like this is not recommended and should be avoided
        // since it can lead to potential security issues.

        // string assemblyQualifiedName = reader.GetString();
        // return Type.GetType(assemblyQualifiedName);
        throw new NotSupportedException();
    }

    public override void Write(Utf8JsonWriter writer, Type value, JsonSerializerOptions options)
    {
        writer.WriteStringValue(value.AssemblyQualifiedName);
    }
}
```

cc @steveharter, @layomia, @Jozkee

### PR DESCRIPTION
**Pull request by layomia**
Fixes https://github.com/dotnet/runtime/issues/31567. Per the conversation in this issue & following triage, this PR adds a new converter for `System.Type` with the following behavior:

- serialization: writes the [`AssemblyQualifiedName`](https://docs.microsoft.com/dotnet/api/system.type.assemblyqualifiedname?view=netcore-3.1) of the `Type` instance. This is compatible with Newtonsoft.Json.

- deserialization: throws `NotSupportedException`, as deserializing `Type` instances from arbitrary user input is a security vulnerability. Relevant path and reader position information is added to the exception message, as applicable.

The behavior before this PR was a `JsonException` (max depth exceeded) on serialization, and `JsonException` on deserialization (most commonly due to the current token being `JsonTokenType.String` rather than `JsonTokenType.StartObject` as the serializer expects when parsing object types).

---

EDIT: the converter will throw a `NotSupportedException` for both serialization and deserialization.

### CODE REVIEW

**Comment by jkotas** [comment]
What's the point of writing `AssemblyQualifiedName` when you cannot deserialize it?

**Comment by jkotas** [comment]
IMHO, we should throw exception for both serialization and deserialization.

**Review by GrabYourPitchforks — CHANGES_REQUESTED** [review]
I don't think we can say _A workaround is to use a custom converter_ and leave it at that. What is the implementation of this theoretical custom converter?

If we believe there's a secure way to implement this, the exception text should point to documentation which shows clearly how to do this safely.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/TypeConverter.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Text.Json.Serialization.Converters
+{
+    internal sealed class TypeConverter : JsonConverter<Type>
+    {
+        public override Type Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            throw ThrowHelper.GetNotSupportedException_SerializeTypeInstanceNotSupported();
```
Comment (jkotas) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/TypeConverter.cs]: These are not performance sensitive methods by any means. You can just throw from here, no need to factor it out into a helper.

**Comment by layomia** [comment]
> IMHO, we should throw exception for both serialization and deserialization.

Makes sense to throw since we won't round-trip otherwise.

**Review by jkotas — APPROVED** [review]

**Comment by layomia** [comment]
> I don't think we can say _A workaround is to use a custom converter_ and leave it at that. What is the implementation of this theoretical custom converter?
> 
> If we believe there's a secure way to implement this, the exception text should point to documentation which shows clearly how to do this safely.

I've removed the note about using a custom converter - users already know they can write one. 

Not sure of a way to implement this safely for arbitrary user input. There doesn't seem to be much appetite to (de)serialize `Type` instances at the moment. If there is in the future, we can add some examples to the JSON docs and extend the exception message to link to them.

**Review by bartonjs — APPROVED** [review]

**Comment by GrabYourPitchforks** [comment]
> Not sure of a way to implement this safely for arbitrary user input.

It's not safe for arbitrary input, but it can be made safe for known-good inputs. For example, maybe your custom converter maintains a static `Dictionary<string, Type>` of known-allowed inputs. When an arbitrary user-provided string comes in, you look up the entry in that static dictionary, either returning the matched `Type` or throwing due to disallowed input. This mechanism could also be used to perform polymorphic deserialization safely.

It really depends on the application's specific scenario.

**Comment by layomia** [comment]
Test failure unrelated - https://github.com/dotnet/runtime/issues/28553.

**Review by GrabYourPitchforks — APPROVED** [review]

**Comment by layomia** [comment]
Test failure unrelated - https://github.com/dotnet/runtime/issues/28553.

**Pull request merged**